spot-contracts/contracts/RolloverVault.sol

```diff
@@ -709,6 +748,46 @@ contract RolloverVault is
     //--------------------------------------------------------------------------
     // Private write methods
 
+    /// @dev Executes a system-level rebalance. This operation transfers value between
+    ///      the perp reserve and the vault such that the system moves toward balance.
+    ///      Settles protocol fees, charged as a percentage of the bi-directional flow in value.
+    ///      Performs some book-keeping to keep track of the vault and perp's assets.
+    function _rebalance(IPerpetualTranche perp_, IERC20Upgradeable underlying_) private {
+        // If the vault has any perp tokens we deconstruct it
+        // such that it's value reflects in the vault's TVL,
+        // before the rebalance calculation.
+        _meldPerps(perp_);
+
+        SubscriptionParams memory s = _querySubscriptionState(perp_);
+        RebalanceData memory r = feePolicy.computeRebalanceData(s);
+        if (r.perpDebasement) {
+            // We transfer value from perp to the vault, by minting the vault perp tokens.
+            uint256 perpSupply = perp_.totalSupply();
+            uint256 perpAmtToVault = perpSupply.mulDiv(r.underlyingAmtToTransfer, s.perpTVL);
+            perp_.debase(perpAmtToVault);
+            if (r.protocolFeeUnderlyingAmt > 0) {
+                // NOTE: We first mint the vault perp tokens, and then pay the protocol fee.
+                uint256 protocolFeePerpAmt = perpSupply.mulDiv(r.protocolFeeUnderlyingAmt, s.perpTVL);
+                perp_.debase(protocolFeePerpAmt);
+                IERC20Upgradeable(address(perp_)).safeTransfer(owner(), protocolFeePerpAmt);
```

Comment:
As it's structured now, the protocol pays fees in spot in one case and ampl in the other. Since the fee is already denominated in ampl, I think it makes sense to keep it simple and just pay in ampl both ways.

We could move one meld operation above, then one conditional ampl transfer after.

Reply:
Yea that is cleaner